grobid-core/src/main/java/org/grobid/core/engines/DateParser.java

```diff
@@ -209,65 +218,76 @@ else if ((date.getDayString().charAt(date.getDayString().length() - 1) == '-')
             for (int n = 0; n < dayString.length(); n++) {
                 char c = dayString.charAt(n);
                 if (Character.isDigit(c)) {
-                    dayStringBis += c;
+                    dayStringBis.append(c);
                 }
             }
             try {
-                int day = Integer.parseInt(dayStringBis);
-                date.setDay(day);
+                int day = Integer.parseInt(dayStringBis.toString());
+                normalizedDate.setDay(day);
             } catch (Exception e) {
                 //e.printStackTrace();
             }
         }
 
         //normalize month
-        if (date.getMonthString() != null) {
+        if (isNotBlank(date.getMonthString())) {
             String month = date.getMonthString().trim();
+            normalizedDate.setMonthString(month);
             int n = 0;
             while (n < 12) {
                 Matcher ma = months[n].matcher(month);
                 if (ma.find()) {
-                    date.setMonth(n + 1);
+                    normalizedDate.setMonth(n + 1);
                     break;
                 }
                 n++;
             }
         }
 
-        if (date.getYearString() != null) {
-            String yearStringBis = "";
+        if (StringUtils.isNotBlank(date.getYearString())) {
+            StringBuilder yearStringBis = new StringBuilder();
             String yearString = date.getYearString().trim();
+            normalizedDate.setYearString(yearString);
             for (int n = 0; n < yearString.length(); n++) {
                 char c = yearString.charAt(n);
                 if (Character.isDigit(c)) {
-                    yearStringBis += c;
+                    yearStringBis.append(c);
                 }
             }
             try {
-                int year = Integer.parseInt(yearStringBis);
+                int year = Integer.parseInt(yearStringBis.toString());
                 if ((year >= 20) && (year < 100)) {
                     year = year + 1900;
                 } else if ((year >= 0) && (year < 20)) {
                     year = year + 2000;
                 }
-                date.setYear(year);
+                normalizedDate.setYear(year);
             } catch (Exception e) {
                 //e.printStackTrace();
             }
         }
 
         // if we don't have day and month, but a year with 8 digits, we might have a YYYYMMDD pattern
-        if (date.getDay() == -1 && date.getMonth() == -1 && date.getYear() != -1 && date.getYear() > 19000000 && date.getYear() < 20251231) {
+        int maxYear = Calendar.getInstance().getWeekYear() + 4;
+        if (date.getDay() == -1 && date.getMonth() == -1 && date.getYear() != -1 && date.getYear() > 19000000 && date.getYear() < maxYear * 10000+1231) {
             int yearPart = date.getYear() / 10000;
-            if (yearPart > 1900 && yearPart < 2025) {
+            if (yearPart > 1900 && yearPart < maxYear) {
```

Comment:
time flies, I guess when that line was written we thought there would be flying cars all over by now.